---
title: "Class_Oct17"
author: "Basanta Khakurel"
date: "`r Sys.Date()`"
output: html_document
---

## Class Notes Oct 18, 2022

# Genetic Data and Mapping 

---
Install `devtools` and some r-open science packages to work with open data as well. 

```{r, eval=FALSE}
install.packages("devtools")
devtools::install_github("ropensci/rotl")
devtools::install_github("ropensci/traits")
devtools::install_github(repo='ropensci/phylotaR', build_vignettes=TRUE)
```

```{r}
library(rotl)
library(phylotaR)
```

Write a vector of 5 organisms. 
`resolved_names` checks the OTL for any spelling mismatches. 
`Induced Subtree` - 
```{r}
apes <- c("Pongo", "Pan", "Gorilla", "Hoolock", "Homo")
(resolved_names <- rotl::tnrs_match_names(apes))

tr_apes <- tol_induced_subtree(ott_ids = ott_id(resolved_names))

plot(tr_apes)
```

```{r}
ants <- c("Martialis", "Atta", "Ectatomma", "Tatuidris", "Aneuretus", "Xymmer", "Aenictus")
(resolved_names <- rotl::tnrs_match_names(ants))

tr_ants <- tol_induced_subtree(ott_ids = ott_id(resolved_names))

plot(tr_ants)
```
---
Looking for genetic data for our taxa of interest.
Generate API key first in the NCBI page. 
```
set_entrez_key("61636659d406b4da66e3388530a1c6cd2b08")
Sys.getenv("ENTREZ_KEY")
```

```{r}
library(rentrez)
entrez_search(db = "Nucleotide", term="Ectatomma", api_key ="61636659d406b4da66e3388530a1c6cd2b08")
```

Narrow down the number of sequences and save it to a variable. 
```{r}
Ectatomma <- entrez_search(db = "Nucleotide", term="Ectatomma AND 2019:2022[PDAT]", retmax = 100, api_key ="61636659d406b4da66e3388530a1c6cd2b08")
```
Run the loop and take a look at the output.

`cat` - goes through each object and writes it out to the specific file.
`seq_start` - generate sequences between two numbers. Could modify to pull out different amount of data if you wanted to. 
```{r}
for( seq_start in seq(1,75,1)){
  recs <- entrez_fetch(db="nuccore", id = Ectatomma$ids,rettype="fasta", retmax=75)
  cat(recs, file="../Ect.seqs.fasta", append=TRUE)
  cat(seq_start+74, "sequences downloaded\r")
}
```
---
--- See if you can modify the entrez_search query to check for a specific gene or constrain the genus to a specific phyla ---
---
```{r}
search_year <- function(ant){
  query <- paste0(ant, " AND ", "2016:2019[PDAT]")
  search_returns <- c(entrez_search(db="Nucleotide", term=query, retmax=20,api_key ="61636659d406b4da66e3388530a1c6cd2b08"))
  return(search_returns)
}
```


ants <- as.vector(resolved_names$unique_name)

labels <- sapply(ants,search_year,  USE.NAMES = T)

search_year <- function(ant){
  search_returns = c()
  query <- paste0(ant, " AND ", "2016:2019[PDAT]")
  search_returns[ant] <- c(entrez_search(db="Nucleotide", term=query, retmax=20,api_key ="61636659d406b4da66e3388530a1c6cd2b08"))
  return(search_returns)
}


ants <- as.vector(resolved_names$unique_name)

labels <- sapply(ants,search_year,  USE.NAMES = T)

xymmer_papers <- entrez_fetch(db="pmc", term="Xymmer", id = xymmer_papers$ids, retmax=20, rettype= "text", api_key ="61636659d406b4da66e3388530a1c6cd2b08")

search_year <- function(ant){
  search_returns = c()
  query <- paste0(ant, " AND ", "HINT HINT HINT:2019[PDAT]")
  search_returns[ant] <- c(entrez_search(db="Nucleotide", term=query, retmax=1,api_key ="61636659d406b4da66e3388530a1c6cd2b08"))
  return(search_returns)
}


ants <- as.vector(resolved_names$unique_name)

labels <- sapply(ants,search_year,  USE.NAMES = F)

search_seq <- function(labels){
  recs[ant] <- entrez_fetch(db="nuccore", id = labels,
                            rettype="fasta", retmax=1)
  cat(recs, file="Ants.seqs.fasta", append=TRUE)
  return(recs)
}


seq_returns <- sapply(labels, search_seq,  USE.NAMES = T)
```
```{r}
install.packages("ape")

coi <- ape::read.dna("Ants.seqs.fasta", format="fasta")
```